src/TBuffer.cpp

```diff
@@ -1784,7 +1804,214 @@ void TBuffer::translateToPlainText( std::string & s )
+            mMudLine.append(QString(decodeByteToWindows_1252(ch)));
+        }
+        else if (mpHost->mTelnet.getEncoding() == QLatin1String("UTF-8")) {
+            // In Utf-8 mode we have to process the data more than one byte at a
+            // time because there is not necessarily a one-byte to one TChar
+            // mapping, instead we use one TChar per QChar - and that has to be
+            // tweaked for non-BMP characters that use TWO QChars per codepoint.
+            if (localBuffer[msPos] & 0x80) {
+                // MSB is set, so if this is Utf-8 then assume this is the first byte
+                size_t utf8SequenceLength = 1;
+                if        ((localBuffer[msPos] & 0xE0) == 0xC0) {
+                    // 2 byte sequence - Unicode code-points: U+00000080 to U+000007FF
+                    utf8SequenceLength = 2;
+                } else if ((localBuffer[msPos] & 0xF0) == 0xE0) {
+                    // 3 byte sequence - Unicode code-points: U+00000800 to U+0000FFFF
+                    utf8SequenceLength = 3;
+                } else if ((localBuffer[msPos] & 0xF8) == 0xF0) {
+                    // 4 byte sequence - Unicode code-points: U+00010000 to U+001FFFFF (<= U+0010FFF LEGAL)
+                    utf8SequenceLength = 4;
+                } else if ((localBuffer[msPos] & 0xFC) == 0xF8) {
+                    // 5 byte sequence - Unicode code-points: U+00200000 to U+03FFFFFF (ALL ILLEGAL)
+                    utf8SequenceLength = 5;
+                } else if ((localBuffer[msPos] & 0xFE) == 0xFC) {
+                    // 6 byte sequence - Unicode code-points: U+04000000 to U+7FFFFFFF (ALL ILLEGAL)
+                    utf8SequenceLength = 6;
+                }
+
+                if (msPos + utf8SequenceLength >= msLength) {
+                    // Not enough bytes left in localBuffer to complete the utf-8
+                    // sequence - need to save and prepend onto incoming data next
+                    // time around.
+                    // The absence of a second argument takes all the available
+                    // bytes - this is only for data from the Server NOT from
+                    // locally generated material from Lua feedTriggers(...)
+                    if( isFromServer ) {
+                        mIncompleteUtf8SequenceBytes = localBuffer.substr(msPos);
+                    }
+                    return; // Bail out
+                }
+
+                // If we have got here we have enough bytes to work with:
+                bool isValid = true;
+                bool isToUseReplacementMark = false;
+                bool isToUseByteOrderMark = false; // When BOM seen in stream it transcodes as zero characters
+                switch( utf8SequenceLength ) {
+                case 4:
+                    if ((localBuffer[msPos+3] & 0xC0) != 0x80) {
+                        qDebug() << "TBuffer::translateToPlainText(...) byte 4 in sequence as UTF-8 rejected!";
```

Comment:
Oh. I thought it literally said 4. The error message still is not very useful: it does not say what the byte was or why was it rejected. How about "4th byte <byte in readable form> in UTF-8 sequence is invalid"